pkg/assembler/backends/inmem/vulnMetadata.go

```diff
@@ -0,0 +1,302 @@
+		return nil, gqlerror.Errorf("uneven vulnerabilities and vulnerabilityMetadatas for ingestion")
+	}
+	var modelVulnMetadataIDList []string
+	for i := range vulnerabilityMetadatas {
+		vulnMetadata, err := c.IngestVulnerabilityMetadata(ctx, *vulnerabilities[i], *vulnerabilityMetadatas[i])
+		if err != nil {
+			return nil, gqlerror.Errorf("IngestVulnerabilityMetadata failed with err: %v", err)
+		}
+		modelVulnMetadataIDList = append(modelVulnMetadataIDList, vulnMetadata)
+	}
+	return modelVulnMetadataIDList, nil
+}
+
+func (c *demoClient) IngestVulnerabilityMetadata(ctx context.Context, vulnerability model.VulnerabilityInputSpec, vulnerabilityMetadata model.VulnerabilityMetadataInputSpec) (string, error) {
+	return c.ingestVulnerabilityMetadata(ctx, vulnerability, vulnerabilityMetadata, true)
+}
+
+func (c *demoClient) ingestVulnerabilityMetadata(ctx context.Context, vulnerability model.VulnerabilityInputSpec, vulnerabilityMetadata model.VulnerabilityMetadataInputSpec, readOnly bool) (string, error) {
+	funcName := "IngestVulnerabilityMetadata"
+	lock(&c.m, readOnly)
+	defer unlock(&c.m, readOnly)
+
+	var vulnerabilityLinks []uint32
+
+	vulnID, err := getVulnerabilityIDFromInput(c, vulnerability)
+	if err != nil {
+		return "", gqlerror.Errorf("%v ::  %s", funcName, err)
+	}
+	foundVulnNode, err := byID[*vulnIDNode](vulnID, c)
+	if err != nil {
+		return "", gqlerror.Errorf("%v ::  %s", funcName, err)
+	}
+	vulnerabilityLinks = foundVulnNode.vulnMetadataLinks
+
+	var searchIDs []uint32 = vulnerabilityLinks
+
+	// Don't insert duplicates
+	duplicate := false
+	collectedVulnMetadataLink := vulnerabilityMetadataLink{}
+	for _, id := range searchIDs {
+		v, err := byID[*vulnerabilityMetadataLink](id, c)
+		if err != nil {
+			return "", gqlerror.Errorf("%v ::  %s", funcName, err)
+		}
+		vulnMatch := false
+		if vulnID != 0 && vulnID == v.vulnerabilityID {
+			vulnMatch = true
+		}
+		if vulnMatch && vulnerabilityMetadata.Timestamp.Equal(v.timestamp) && vulnerabilityMetadata.ScoreType == v.scoreType &&
+			vulnerabilityMetadata.ScoreValue == float64(v.scoreValue) &&
+			vulnerabilityMetadata.Origin == v.origin && vulnerabilityMetadata.Collector == v.collector {
+
+			collectedVulnMetadataLink = *v
+			duplicate = true
+			break
+		}
+	}
+
+	if !duplicate {
+		if readOnly {
+			c.m.RUnlock()
+			cv, err := c.ingestVulnerabilityMetadata(ctx, vulnerability, vulnerabilityMetadata, false)
+			c.m.RLock() // relock so that defer unlock does not panic
+			return cv, err
+		}
+		// store the link
+		collectedVulnMetadataLink = vulnerabilityMetadataLink{
+			id:              c.getNextID(),
+			vulnerabilityID: vulnID,
+			timestamp:       vulnerabilityMetadata.Timestamp,
+			scoreType:       vulnerabilityMetadata.ScoreType,
+			scoreValue:      (vulnerabilityMetadata.ScoreValue),
+			origin:          vulnerabilityMetadata.Origin,
+			collector:       vulnerabilityMetadata.Collector,
+		}
+		c.index[collectedVulnMetadataLink.id] = &collectedVulnMetadataLink
```

Comment:
That said, structs are usually cleaner as pointers, so:
```
var collectedVulnMetadataLink *vulnerabilityMetadataLink
...
collectedVulnMetadataLink = v
...
collectedVulnMetadataLink = &vulnerabilityMetadataLink{
...
c.index[collectedVulnMetadataLink.id] = collectedVulnMetadataLink
```
looks nicer.

Reply:
fixed!